sdk/js-query/src/query/BinaryWriter.ts

```diff
@@ -0,0 +1,54 @@
+// BinaryWriter appends data to the end of a buffer, resizing the buffer as needed
+// Numbers are encoded as big endian
+export class BinaryWriter {
+  private _buffer: Buffer;
+  private _offset: number;
+
+  constructor(initialSize: number = 1024) {
+    this._buffer = Buffer.alloc(initialSize);
+    this._offset = 0;
+  }
+
+  // Ensure the buffer has the capacity to write `size` bytes, otherwise allocate more memory
+  _ensure(size: number) {
+    const remaining = this._buffer.length - this._offset;
+    if (remaining < size) {
+      const oldBuffer = this._buffer;
+      const newSize = this._buffer.length * 2 + size;
+      this._buffer = Buffer.alloc(newSize);
+      oldBuffer.copy(this._buffer);
+    }
+  }
+
+  writeUint8(value: number) {
+    this._ensure(1);
+    this._buffer.writeUint8(value, this._offset);
```

Comment:
Node docs for Buffer have a scary-ish:
> Behavior is undefined when value is anything other than an unsigned 8-bit integer.

Maybe a check that the value is between 0 and 2**8-1 ?

Reply:
Good idea